metricbeat/module/consul/agent/agent.go

```
@@ -0,0 +1,99 @@
+	"github.com/elastic/beats/metricbeat/mb"
+	"github.com/elastic/beats/metricbeat/mb/parse"
+)
+
+var (
+	hostParser = parse.URLHostParserBuilder{
+		DefaultScheme: "http",
+		DefaultPath:   "/v1/agent/metrics",
+	}.Build()
+)
+
+// init registers the MetricSet with the central registry as soon as the program
+// starts. The New function will be called later to instantiate an instance of
+// the MetricSet for each host defined in the module's configuration. After the
+// MetricSet has been created then Fetch will begin to be called periodically.
+func init() {
+	mb.Registry.MustAddMetricSet("consul", "agent", New,
+		mb.WithHostParser(hostParser))
+}
+
+// MetricSet holds any configuration or state information. It must implement
+// the mb.MetricSet interface. And this is best achieved by embedding
+// mb.BaseMetricSet because it implements all of the required mb.MetricSet
+// interface methods except for Fetch.
+type MetricSet struct {
+	mb.BaseMetricSet
+	http *helper.HTTP
+}
+
+// New creates a new instance of the MetricSet. New is responsible for unpacking
+// any MetricSet specific configuration options if there are any.
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Beta("The consul agent metricset is beta.")
+
+	config := struct {
+		Schema string   `config:"schema"`
+		Hosts  []string `config:"hosts"`
+	}{}
+	if err := base.Module().UnpackConfig(&config); err != nil {
+		return nil, err
+	}
+
+	http, err := helper.NewHTTP(base)
+	if err != nil {
+		return nil, err
+	}
+
+	return &MetricSet{
+		BaseMetricSet: base,
+		http:          http,
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right
+// format. It publishes the event which is then forwarded to the output. In case
+// of an error set the Error field of mb.Event or simply call report.Error().
+func (m *MetricSet) Fetch(report mb.ReporterV2) {
+	content, err := m.http.FetchContent()
+	if err != nil {
+		logp.Error(err)
```

We should also report an error here

---

Just curious. What should be the general approach? log and report or only report? (added report)

---

In most cases we should report and error and log it too. Currently the code to do this looks ugly and we should come up with something simpler in the future.

The reason to log in addition to sending the event is like this also the operator sees that something is wrong in case it's not the same person that consumes the data.

---

Error now is also reported.

---

Related #9091